### PR TITLE
Improvments in handling TLS cipher suite

### PR DESCRIPTION
**Pull request by ryru**
This PR addresses these three issues:

- #5 Review cipher suite order
- #6 Do not suggest TLS 1.3 when no valid cipher suite is available
- #7 Don't HelloRequest with list of 0 cipher suites

Review und feedback is very welcome!

### CODE REVIEW

File: src/libtls/tls_crypto.c [hunk 1]
```diff
@@ -518,74 +518,86 @@ static suite_algs_t suite_algs[] = {
...
-		KEY_ECDSA, ECP_256_BIT,
+	{ TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+		KEY_RSA, ECP_256_BIT,
 		HASH_SHA256, PRF_HMAC_SHA2_256,
-		AUTH_HMAC_SHA2_256_256, ENCR_AES_CBC, 16,
+		AUTH_UNDEFINED, ENCR_AES_GCM_ICV16, 16,
 		TLS_1_2, TLS_1_2,
 	},
-	{ TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
+	{ TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c:533]: Any particular reason why the 256-bit cipher suites are after the 128-bit suites?

File: src/libtls/tls_crypto.c [hunk 2]
```diff
@@ -518,74 +518,86 @@ static suite_algs_t suite_algs[] = {
...
-	{ TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+	{ TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256,
 		KEY_ECDSA, ECP_256_BIT,
 		HASH_SHA256, PRF_HMAC_SHA2_256,
-		AUTH_UNDEFINED, ENCR_AES_GCM_ICV16, 16,
+		AUTH_HMAC_SHA2_256_256, ENCR_AES_CBC, 16,
 		TLS_1_2, TLS_1_2,
 	},
-	{ TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+	{ TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA,
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c:545]: Same question as above.

File: src/libtls/tls_crypto.c [hunk 3]
```diff
@@ -518,74 +518,86 @@ static suite_algs_t suite_algs[] = {
...
-		KEY_RSA, ECP_256_BIT,
+	{ TLS_DHE_RSA_WITH_AES_128_GCM_SHA256,
+		KEY_RSA, MODP_3072_BIT,
 		HASH_SHA256, PRF_HMAC_SHA2_256,
 		AUTH_UNDEFINED, ENCR_AES_GCM_ICV16, 16,
 		TLS_1_2, TLS_1_2,
 	},
-	{ TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
-		KEY_RSA, ECP_384_BIT,
+	{ TLS_DHE_RSA_WITH_AES_256_GCM_SHA384,
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c:605]: Same here, although the overhead of modp4096 might be an argument in this case why the 128-bit suite might be preferable.

File: src/libtls/tls_crypto.c [hunk 4]
```diff
@@ -650,46 +650,46 @@ static suite_algs_t suite_algs[] = {
...
-	},
 	{ TLS_RSA_WITH_AES_256_CBC_SHA,
 		KEY_RSA, MODP_NONE,
 		HASH_SHA256, PRF_HMAC_SHA2_256,
 		AUTH_HMAC_SHA1_160, ENCR_AES_CBC, 32,
 		SSL_3_0, TLS_1_2,
 	},
-	{ TLS_RSA_WITH_AES_256_CBC_SHA256,
-		KEY_RSA, MODP_NONE,
+	{ TLS_DHE_RSA_WITH_3DES_EDE_CBC_SHA,
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c]: I guess we could also remove all the 3-DES suites. And I also wonder about the suites without DH exchange. Maybe something to discuss on Wednesday.

File: src/libtls/tls_peer.c [hunk 5]
```diff
@@ -1217,6 +1217,7 @@ static status_t send_client_hello(private_tls_peer_t *this,
 	}
 	rng->destroy(rng);
 
+	count = this->crypto->get_cipher_suites(this->crypto, &suites);
```
Comment (tobiasbrunner) [inline src/libtls/tls_peer.c:1221]: Add a comment and perhaps a newline.

```suggestion
	/* determine supported suites before the versions as they might change */
	count = this->crypto->get_cipher_suites(this->crypto, &suites);
	
```

File: src/libtls/tls_crypto.c [hunk 6]
```diff
@@ -1198,11 +1198,30 @@ static void build_cipher_suite_list(private_tls_crypto_t *this)
...
 	{
 		DBG2(DBG_TLS, "  %N", tls_cipher_suite_names, suites[i].suite);
 		this->suites[i] = suites[i].suite;
+
+		/* set TLS min/max versions appropriate to the final cipher suites */
+		if (new_max_version < suites[i].max_version &&
+			max_version >= suites[i].max_version)
+		{
+			new_max_version = suites[i].max_version;
+		}
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c]:
```suggestion
		new_max_version = max(new_max_version, suites[i].max_version);
```

File: src/libtls/tls_crypto.c [hunk 7]
```diff
@@ -1198,11 +1198,30 @@ static void build_cipher_suite_list(private_tls_crypto_t *this)
...
+		if (new_min_version > suites[i].min_version &&
+			min_version <= suites[i].min_version)
+		{
+			new_min_version = suites[i].min_version;
+		}
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c]:
```suggestion
		new_min_version = min(new_min_version, suites[i].min_version);
```

File: src/libtls/tls_crypto.c [hunk 8]
```diff
@@ -1198,11 +1198,30 @@ static void build_cipher_suite_list(private_tls_crypto_t *this)
...
 	}
+
+	this->tls->set_version(this->tls, new_min_version, new_max_version);
+	DBG2(DBG_TLS, "TLS min/max %N/%N according to the cipher suites",
+	  	 tls_version_names, new_min_version,
+	  	 tls_version_names, new_max_version);
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c]: I'd make this conditional on an actual difference between the versions (i.e. add something like `if (min_version != new_min_version || max_version != new_max_version)`).

File: src/libtls/tls_crypto.c [hunk 9]
```diff
@@ -1198,11 +1198,30 @@ static void build_cipher_suite_list(private_tls_crypto_t *this)
...
+		{
+			new_max_version = suites[i].max_version;
+		}
+		if (new_min_version > suites[i].min_version &&
+			min_version <= suites[i].min_version)
+		{
+			new_min_version = suites[i].min_version;
+		}
 	}
+
```
Comment (tobiasbrunner) [inline src/libtls/tls_crypto.c:1184]:
```suggestion
	new_max_version = min(new_max_version, max_version);
	new_min_version = max(new_min_version, min_version);
```

File: src/libtls/tls_peer.c [hunk 10]
```diff
@@ -1237,6 +1237,11 @@ static status_t send_client_hello(private_tls_peer_t *this,
 
 	/* add TLS cipher suites */
 	count = this->crypto->get_cipher_suites(this->crypto, &suites);
+	if (count <= 0)
+	{
+		DBG1(DBG_TLS, "no supported TLS cipher suite available");
+		return FAILED;
```
Comment (tobiasbrunner) [inline src/libtls/tls_peer.c]: Why not add a `TLS_INTERNAL_ERROR` alert and return `NEED_MORE`? (That's what's done in the other two error cases in this function.) I guess the changes in tls-socket wouldn't be necessary then.

**Review by tobiasbrunner — COMMENTED** [review]
Please don't end the first line of the commit message with a dot. You could also use more specific component identifiers (e.g. `tls-crypto:` for the first commit).

File: src/libtls/tls_peer.c (hunk 10 above)
Comment (ryru) [reply src/libtls/tls_peer.c]: I agree.

File: src/libtls/tls_crypto.c (hunk 6 above)
Comment (ryru) [reply src/libtls/tls_crypto.c]: I agree.

File: src/libtls/tls_crypto.c (hunk 7 above)
Comment (ryru) [reply src/libtls/tls_crypto.c]: I agree.

File: src/libtls/tls_crypto.c (hunk 9 above)
Comment (ryru) [reply src/libtls/tls_crypto.c:1184]: I agree.

File: src/libtls/tls_crypto.c (hunk 8 above)
Comment (ryru) [reply src/libtls/tls_crypto.c]: I agree.

File: src/libtls/tls_peer.c (hunk 5 above)
Comment (ryru) [reply src/libtls/tls_peer.c:1221]: I agree.

File: src/libtls/tls_crypto.c (hunk 1 above)
Comment (ryru) [reply src/libtls/tls_crypto.c:533]: This order is inspired by Firefox 80.0.1. Relying on this makes sense to me. Although a descending approach according to the following rule set would be good to me as well:

0. TLS 1.3 > Legacy TLS
1. AES > CAMELLIA > 3DES > NULL
2. AES256 > AES128
3. ECDHE > DHE > NULL
4. ECDSA > RSA
5. SHA384 > SHA256 > SHA1 > MD5

What do you think?

File: src/libtls/tls_crypto.c (hunk 4 above)
Comment (ryru) [reply src/libtls/tls_crypto.c]: I agree. Let's discuss on Wednesday.

File: src/libtls/tls_crypto.c (hunk 1 above)
Comment (ryru) [reply src/libtls/tls_crypto.c:533]: In today's meeting we agreed on the above rule set. All cipher suites shall be ordered accordingly whereby we remove ciphers with 3DES and/or MD5.

File: src/libtls/tls_crypto.c (hunk 4 above)
Comment (ryru) [reply src/libtls/tls_crypto.c]: We agreed on removing 3DES and MD5 suites. We have not particularly discussed what to do with suites without DH exchange. I'll keep them for now, maybe remove them later.

**Pull request merged**